### PR TITLE
[PATCH] Webpack 4 compatibility work

### PR DESCRIPTION
**Pull request by maxnachlinger**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -14,45 +15,46 @@ function needsArray(value) {
...
 
-module.exports = function(source) {
-  var self = this;
-  if (this.cacheable) {
-    this.cacheable();
-  }
-  var done = this.async();
-  var options = loaderUtils.parseQuery(this.query);
+function calcOptions(source, initialOptions, lib) {
+  var options = cloneDeep(initialOptions);
```
Comment (maxnachlinger) [inline index.js]: Make sure we don't mutate what we down own

File: index.js [hunk 2]
```diff
@@ -20,26 +20,35 @@ module.exports = function(source) {
     this.cacheable();
   }
   var done = this.async();
-  var options = loaderUtils.parseQuery(this.query);
+  var options = cloneDeep(loaderUtils.parseQuery(this.query) || {});
```
Comment (maxnachlinger) [inline index.js]: Makes sure we don't mutate what we don't own.

File: index.js [hunk 3]
```diff
@@ -20,26 +20,35 @@ module.exports = function(source) {
...
+
   options.dest = options.dest || '';
   options.filename = options.filename || this.resourcePath;
   options.Evaluator = CachedPathEvaluator;
-
   // Attach `importCache` to `options` so that the `Evaluator` can access it.
   var importCache = options.importCache = new ImportCache(this, options);
```
Comment (maxnachlinger) [inline index.js:29]: Webpack 4 doesn’t provide `this.options`, see: https://github.com/shama/stylus-loader/commit/fc81e32e684dc95832d5ef0f24703265d1e132f6

File: index.js [hunk 4]
```diff
@@ -49,10 +58,8 @@ module.exports = function(source) {
   }
 
   var styl = stylus(source, options);
-  var paths = [path.dirname(options.filename)];
```
Comment (maxnachlinger) [inline index.js:52]: `paths` didn't seem to be used :)

**Comment by maxnachlinger** [comment]
Sadly the tests seem to be failing on master as well (at least on my machine)

File: index.js [hunk 5]
```diff
@@ -1,7 +1,6 @@
 'use strict';
 var loaderUtils = require('loader-utils');
 var stylus = require('stylus');
-var path = require('path');
```
Comment (maxnachlinger) [inline index.js:4]: `path` was only used to build-up the `paths` variable, and the `paths` variable isn't used.

**Comment by maxnachlinger** [comment]
Here's an example of how I'm using this branch with webpack 4:
```js
const MiniCssExtractPlugin = require('mini-css-extract-plugin');

module.exports = {
  // snip
  module: {
    rules: [
      // snip
      {
        test: /\.styl$/,
        use: [
          MiniCssExtractPlugin.loader,
          'css-loader',
          'stylus-relative-loader'
        ],
      },
      // snip
    ]
  },
  resolve: {
    extensions: ['.js', '.jsx', '.styl'],
    // snip
  },
  plugins: [
    // snip
    new MiniCssExtractPlugin({
      filename: 'bundle.css'
    })
    // snip
  ]
};
```

**Comment by jchip** [comment]
are these changes still compatible with webpack 3 or a major is required?

**Comment by maxnachlinger** [comment]
@jchip - not sure re: webpack 3, I was thinking of making an `/examples` folder and throwing in an example app using webpack 2, 3 and 4. Perhaps if I do that we'll know :) I've only tested webpack 1 and 4 so far and these changes work with both.

**Comment by ryan-roemer** [comment]
Hi folks! In case it helps you definitely can write webpack1-4 compatible plugins -- here are some relevant different snippets from two of ours that are fully tested in all of webpack1-4:

- Simple: https://github.com/FormidableLabs/webpack-stats-plugin/blob/master/lib/stats-writer-plugin.js#L52-L66
- Wrapped and complex: https://github.com/FormidableLabs/webpack-dashboard/blob/master/plugin/index.js#L37-L64

Good luck!

**Pull request merged**